In [8]:
import sys, io
import pandas as pd
import plotly.express as px
from fractions import Fraction

pd.set_option('display.max_rows', None)

In [31]:
def CollatzChain(collatzNumber):
    chain = [collatzNumber]
    while collatzNumber != 1:
        if collatzNumber & 1 == 0:
            collatzNumber = collatzNumber // 2
        else:
            collatzNumber = (3 * collatzNumber + 1) // 2
        chain.append(collatzNumber)
    return chain
#

def ChainPath(collatzNumber):
    path = []
    while collatzNumber != 1:
        if (collatzNumber & 1) == 0:
            collatzNumber = collatzNumber // 2
            path.append("1")
        else:
            collatzNumber = (3 * collatzNumber + 1) // 2
            path.append("0")
    return "".join(path)
#
def fractionFromNodeTup(tup):
    p2, p3, c = tup
    fract = Fraction(2**p2 - c, 3**p3)
    return (fract.numerator, fract.denominator)
#

def TupChainFromPath(chain_path):
    tup_chain = [(0, 0, 0)]
    for chain_item in chain_path:
        p2, p3, c = tup_chain[-1]
        if chain_item == "1":
            tup_chain. append((p2 + 1, p3, c))
        else:
            tup_chain. append((p2 + 1, p3 + 1, c*3 + 2**p2))
            
        fract = fractionFromNodeTup(tup_chain[-1])
    return tup_chain
#
def FractionFromPath(chain_path):
    tup = (0, 0, 0)
    for chain_item in chain_path:
        p2, p3, c = tup
        if chain_item == "1":
            tup = (p2 + 1, p3, c)
        else:
            tup = (p2 + 1, p3 + 1, c*3 + 2**p2)
        
    fract = fractionFromNodeTup(tup)
    return fract
#    
def downUpTup(tup):
    p2, p3, c = tup
    p2_01, p3_01, c_01 = (p2 + 2, p3 + 1, c*3 + 2**p2)
    return (p2_01, p3_01, c_01)
#
def TupChain(collatzNumber):
    chain_path = ChainPath(collatzNumber)
    return TupChainFromPath(chain_path)
    return tup_chain
#
def generationTups(n):
    G = [[(0, 0, 0)]]
    for i in range(1, n+1, 1):
        G.append([])
        tups = G[i-1]
        for tup in tups:
            p2, p3, c = tup
            G[i].extend([(p2 + 1, p3, c), (p2 + 1, p3 + 1, c*3 + 2**p2)])
    return G
#
def pathToTup(path):
    a = len(path)
    zeros = [i for i, b in enumerate(path) if b == '0']
    b = len(zeros)
    # compute c = sum_{j=1}^{k} 3^{k-j} * 2^{i_j - 1}
    c = sum((3 ** (b - j - 1)) * (2 ** (i)) for j, i in enumerate(zeros))
    fract = Fraction(2**a - c, 3**b)
    return path, (a, b, c), (fract.numerator, fract.denominator)
#
def reverseColllatz(collatz_num, generations):
    chains = [[collatz_num]]
    for i in range(generations):
        for chain in chains.copy():
            if ((chain[-1] -1) % 3) == 0:
                reverse_down_val = (chain[-1] - 1)//3
                if (reverse_down_val & 1) == 1:
                    # down only applies to odds
                    chain2 = chain.copy()
                    chain2.append(reverse_down_val)
                    chains.append(chain2)
            chain = chain.append(2*chain[-1])
    #
    return chains
#


def collatzNext(x):
    if x %2 == 0:
        return x//2
    return 3*x + 1
#
def fourPlusOneEdges(maxN):
    for n in range(maxN):
        start = 4*n + 1
        y = collatzNext(start)
        while y % 4 != 1:
            y = collatzNext(y)
        print((n, (y-1)//4))
#
def ToNPlusOneEdges(g, maxN):
    L = []
    multiplier = 2**g
    for n in range(maxN):
        start = multiplier*n + 1
        y = collatzNext(start)
        while y % multiplier != 1:
            y = collatzNext(y)
        L.append((n, (y-1)//multiplier))
    return L
#


def two_factors(n):
    """
    Returns the exponent of the largest power of 2 that divides n.
    For example, if n = 48 (binary 110000), the result is 4 because 2^4 = 16 divides 48.
    """
    if n <= 1:
        return 0 

    count = 0
    while (n & 1) == 0:
        n >>= 1
        count += 1
    return count
#


In [10]:
L4 = list(zip(ToNPlusOneEdges(2, 100), ToNPlusOneEdges(3, 100), ToNPlusOneEdges(4, 100), ToNPlusOneEdges(5, 100)))
L4

[((0, 0), (0, 0), (0, 0), (0, 0)),
 ((1, 0), (1, 2), (1, 0), (1, 0)),
 ((2, 4), (2, 0), (2, 1), (2, 0)),
 ((3, 1), (3, 2), (3, 1), (3, 5)),
 ((4, 3), (4, 3), (4, 3), (4, 3)),
 ((5, 0), (5, 20), (5, 0), (5, 18)),
 ((6, 7), (6, 2), (6, 10), (6, 5)),
 ((7, 4), (7, 8), (7, 0), (7, 0)),
 ((8, 6), (8, 6), (8, 6), (8, 6)),
 ((9, 4), (9, 20), (9, 10), (9, 0)),
 ((10, 40), (10, 0), (10, 37), (10, 0)),
 ((11, 4), (11, 2), (11, 1), (11, 14)),
 ((12, 9), (12, 9), (12, 9), (12, 9)),
 ((13, 1), (13, 2), (13, 1), (13, 11)),
 ((14, 16), (14, 0), (14, 91), (14, 18)),
 ((15, 13), (15, 17), (15, 1), (15, 2)),
 ((16, 12), (16, 12), (16, 12), (16, 12)),
 ((17, 3), (17, 29), (17, 1), (17, 2)),
 ((18, 31), (18, 5), (18, 0), (18, 0)),
 ((19, 7), (19, 8), (19, 4), (19, 5)),
 ((20, 15), (20, 15), (20, 15), (20, 15)),
 ((21, 0), (21, 182), (21, 37), (21, 20)),
 ((22, 25), (22, 3), (22, 28), (22, 0)),
 ((23, 13), (23, 26), (23, 0), (23, 74)),
 ((24, 18), (24, 18), (24, 18), (24, 18)),
 ((25, 7), (25, 0), (25, 7),

In [11]:
def allSame(L):
    LL = []
    for tup in L:
        all_same = True
        for i in range(1, len(tup), 1):
            if tup[i] != tup[0]:
                all_same = False
                break
        if all_same:
            LL.append(tup[0])
    return LL
#

In [12]:
L4_same = allSame(L4)
L4_same

[(0, 0),
 (4, 3),
 (8, 6),
 (12, 9),
 (16, 12),
 (20, 15),
 (24, 18),
 (28, 21),
 (32, 24),
 (36, 27),
 (40, 30),
 (44, 33),
 (48, 36),
 (52, 39),
 (56, 42),
 (60, 45),
 (64, 48),
 (68, 51),
 (72, 54),
 (76, 57),
 (80, 60),
 (84, 63),
 (88, 66),
 (92, 69),
 (96, 72)]

# (4,3)

    - 17 -> 13:
    - 33 -> 25:
    - 65 -> 49:

In [13]:
for tup in L4_same:
    for i in range(2,6,1):
        val_0 = (2**i) * tup[0] + 1
        val_1 = (2**i) * tup[1] + 1
        print(ChainPath(val_0) + " -> " + ChainPath(val_1))
    

 -> 
 -> 
 -> 
 -> 
010110111 -> 0110111
010100110010110111 -> 0100110010110111
0101011100010110111 -> 01011100010110111
01010100011000010100100010000101100010010000001100001110101011101100011110111 -> 010100011000010100100010000101100010010000001100001110101011101100011110111
010100110010110111 -> 0100110010110111
0101011100010110111 -> 01011100010110111
01010100011000010100100010000101100010010000001100001110101011101100011110111 -> 010100011000010100100010000101100010010000001100001110101011101100011110111
010101011010000010100100010000101100010010000001100001110101011101100011110111 -> 0101011010000010100100010000101100010010000001100001110101011101100011110111
01011100010110111 -> 011100010110111
010100011000010100100010000101100010010000001100001110101011101100011110111 -> 0100011000010100100010000101100010010000001100001110101011101100011110111
0101011010000010100100010000101100010010000001100001110101011101100011110111 -> 01011010000010100100010000101100010010000001100001110101

In [14]:
FractionFromPath("0100110010110111")

(25, 1)

In [15]:
FractionFromPath("010100110010110111")

(33, 1)

In [16]:
FractionFromPath("010011001011011101")

(25, 1)

In [17]:
FractionFromPath("110111")

(20, 1)

In [18]:
for tup in L4_same:
    print("Same %s ____"%(str(tup)))
    for i in range(2,6,1):
        val_0 = (2**i) * tup[0] + 1
        val_1 = (2**i) * tup[1] + 1
        print("%d -> %d"%(val_0, val_1))

Same (0, 0) ____
1 -> 1
1 -> 1
1 -> 1
1 -> 1
Same (4, 3) ____
17 -> 13
33 -> 25
65 -> 49
129 -> 97
Same (8, 6) ____
33 -> 25
65 -> 49
129 -> 97
257 -> 193
Same (12, 9) ____
49 -> 37
97 -> 73
193 -> 145
385 -> 289
Same (16, 12) ____
65 -> 49
129 -> 97
257 -> 193
513 -> 385
Same (20, 15) ____
81 -> 61
161 -> 121
321 -> 241
641 -> 481
Same (24, 18) ____
97 -> 73
193 -> 145
385 -> 289
769 -> 577
Same (28, 21) ____
113 -> 85
225 -> 169
449 -> 337
897 -> 673
Same (32, 24) ____
129 -> 97
257 -> 193
513 -> 385
1025 -> 769
Same (36, 27) ____
145 -> 109
289 -> 217
577 -> 433
1153 -> 865
Same (40, 30) ____
161 -> 121
321 -> 241
641 -> 481
1281 -> 961
Same (44, 33) ____
177 -> 133
353 -> 265
705 -> 529
1409 -> 1057
Same (48, 36) ____
193 -> 145
385 -> 289
769 -> 577
1537 -> 1153
Same (52, 39) ____
209 -> 157
417 -> 313
833 -> 625
1665 -> 1249
Same (56, 42) ____
225 -> 169
449 -> 337
897 -> 673
1793 -> 1345
Same (60, 45) ____
241 -> 181
481 -> 361
961 -> 721
1921 -> 1441
Same (64, 48) ____
257 -> 1

In [30]:
cnum = 5
for i in range(8):
    print(cnum, CollatzChain(cnum), ChainPath(cnum))
    cnum = 4*cnum + 1

5 [5, 8, 4, 2, 1] 0111
21 [21, 32, 16, 8, 4, 2, 1] 011111
85 [85, 128, 64, 32, 16, 8, 4, 2, 1] 01111111
341 [341, 512, 256, 128, 64, 32, 16, 8, 4, 2, 1] 0111111111
1365 [1365, 2048, 1024, 512, 256, 128, 64, 32, 16, 8, 4, 2, 1] 011111111111
5461 [5461, 8192, 4096, 2048, 1024, 512, 256, 128, 64, 32, 16, 8, 4, 2, 1] 01111111111111
21845 [21845, 32768, 16384, 8192, 4096, 2048, 1024, 512, 256, 128, 64, 32, 16, 8, 4, 2, 1] 0111111111111111
87381 [87381, 131072, 65536, 32768, 16384, 8192, 4096, 2048, 1024, 512, 256, 128, 64, 32, 16, 8, 4, 2, 1] 011111111111111111


# Lets start simple:

We can trivially constructuct the lattice paths for all $4n+1$ numbers of the following form by adding "1" to the end of the path for 5, therefore we know these values are are in the lattice

$$
\large \frac{4(4^{i}-1)}{3} +1
$$

This gives us an "anchor" 4n+1 number in every generation in the very middle (top of the bottom half) of the lattice

What odd numbers immediately feed into these numbers?

Markdown ... not working

cnum = 5
for i in range(8):
    prefix = "01"
    fract = FractionFromPath(prefix + ChainPath(cnum))
    while fract[1] != 1:
        print(fract)
        prefix = prefix + "1"
        fract = FractionFromPath(prefix + ChainPath(cnum))    
    print((fract[0], prefix))
    cnum = 4*cnum + 1

19, 83, 113, 1363, 5459, 7281, 87379, ...

In [94]:
# Prepending 1's doubles the even generations, but only the numerator of the odd generations !
five_path = "0111"
path = five_path
for i in range(10):
    f = FractionFromPath("1"+path)
    print(f)
    path = path+"1"

print("-----------------------")
# The generations prepend two ones to get 
path = five_path
for i in range(10):
    f = FractionFromPath("11"+path)
    print(f)
    path = path+"1"

#  I "know" the even generations are different than the odd, but was not expecting THIS difference.

(10, 1)
(62, 3)
(42, 1)
(254, 3)
(170, 1)
(1022, 3)
(682, 1)
(4094, 3)
(2730, 1)
(16382, 3)
-----------------------
(20, 1)
(124, 3)
(84, 1)
(508, 3)
(340, 1)
(2044, 3)
(1364, 1)
(8188, 3)
(5460, 1)
(32764, 3)


In [75]:
for i in range(8):
    parent =  (4*((4**i) -1) / 3) + 1
    val = (2*parent - 1)/3
    print(val)

0.3333333333333333
3.0
13.666666666666666
56.333333333333336
227.0
909.6666666666666
3640.3333333333335
14563.0


In [65]:
cnum = 5
for i in range(8):
    print(cnum, cnum/(3**len(ChainPath(cnum))))
    cnum = 4*cnum + 1

5 0.06172839506172839
21 0.02880658436213992
85 0.01295534217344917
341 0.0057748649426747276
1365 0.0025684883176119266
5461 0.0011417594385412073
21845 0.0005074718699247732
87381 0.0002255456344746909


# Just above the $\frac{4(4^{i}-1)}{3} +1$ numbers

These fractions will always be in the lattice just above the $\large{ \frac{4(4^{i}-1)}{3} +1}$ numbers above:

In [99]:
not_five_path = "1000"
path = not_five_path
for i in range(20):
    f = FractionFromPath(path)
    print((f, f[0]/f[1]))
    path = path+"0"

# The assymtope to -2 is interesting

((-22, 27), -0.8148148148148148)
((-98, 81), -1.2098765432098766)
((-358, 243), -1.4732510288065843)
((-1202, 729), -1.6488340192043895)
((-3862, 2187), -1.7658893461362597)
((-12098, 6561), -1.8439262307575064)
((-37318, 19683), -1.8959508205050044)
((-114002, 59049), -1.9306338803366696)
((-346102, 177147), -1.9537559202244463)
((-1046498, 531441), -1.9691706134829643)
((-3155878, 1594323), -1.9794470756553095)
((-9500402, 4782969), -1.986298050436873)
((-28566742, 14348907), -1.9908653669579153)
((-85831298, 43046721), -1.9939102446386103)
((-257756038, 129140163), -1.9959401630924067)
((-773792402, 387420489), -1.9972934420616046)
((-2322425782, 1162261467), -1.9981956280410698)
((-6969374498, 3486784401), -1.998797085360713)
((-20912317798, 10460353203), -1.9991980569071421)
((-62745342002, 31381059609), -1.999465371271428)


=============================================================================

TODO:  This got seperated from the code above that led to this statement.

ALSO:  This statement is still too ambitious

# So all odd integers of form $2^{3+i}n + 1;\ i,n {\in} \mathbb{Z}^{+}$ are in the Collatz lattice

## Because these numbers always link to a smaller number with a shorter Collatz chain in the same set

The Collatz chains of these numbers:
- Always start with "01"
- Always end with "11011\[1\]+"
- Pass through the Collatz numbers 5,21,85,... on the way to 1
- To get to the next number:
    1. Remove the leading '01'
    2. Remove any additional leading ones
    3. Gives a chain at least 2 shorter than previous

### Outstanding question
- ? Do I need to prove that the zero after removing the leading ones is always followed immediately by a 1? (This is effectively how we have defined the 4n+1 graph ... implicit in its design is the next number is a 4n+1 number
 
# $\frac{1}{4} + \frac{1}{8} + \frac{1}{16} +  ...\ =\ \frac{1}{2}$
!!! This is half of all $4n+1$ integers which is $\frac{1}{4}$ of all odd integers must be Collatz integers

And so would have to be absent the chain of any odd number that was not a Collatz number


In [20]:
ChainPath(5)

'0111'

In [21]:
FractionFromPath("11110110111")

(208, 1)

In [22]:
# Remove leading ones from ^^^ and we get next $4n+1$ number
FractionFromPath("0110111")

(13, 1)

In [23]:
CollatzChain(25)

[25, 38, 19, 29, 44, 22, 11, 17, 26, 13, 20, 10, 5, 8, 4, 2, 1]

In [24]:
# 4n+1 numbers => 3n+1 numbers which are also 4m+1 numbers 1/4 of the time.
# case "010...110111 "
for n in range(1, 40, 1):
    x = 4*n + 1
    y = (3*x + 1)//4
    if y % 4 == 1:
        print((n, x, "%4->", y))

(4, 17, '%4->', 13)
(8, 33, '%4->', 25)
(12, 49, '%4->', 37)
(16, 65, '%4->', 49)
(20, 81, '%4->', 61)
(24, 97, '%4->', 73)
(28, 113, '%4->', 85)
(32, 129, '%4->', 97)
(36, 145, '%4->', 109)


In [25]:
# 8n+1 numbers => 3n+1 numbers which are also 8m+1 numbers 1/8 of the time.
# case "0110...110111 "
for n in range(1, 40, 1):
    x = 8*n + 1
    y = (3*x + 1)//8
    if y % 8 == 1:
        print((n, x, "%8->", y))
    elif y % 4 == 1:
        print((n, x, "%4->", y))        

(3, 25, '%8->', 9)
(7, 57, '%4->', 21)
(11, 89, '%8->', 33)
(15, 121, '%4->', 45)
(19, 153, '%8->', 57)
(23, 185, '%4->', 69)
(27, 217, '%8->', 81)
(31, 249, '%4->', 93)
(35, 281, '%8->', 105)
(39, 313, '%4->', 117)


In [26]:
# 16n+1 numbers => 3n+1 numbers which are also 16m+1 numbers 1/16 of the time.
# case "01110...110111 "
for n in range(1, 80, 1):
    x = 16*n + 1
    y = (3*x + 1)//16
    if y % 16 == 1:
        print((n, x, "%16->", y))
    elif y % 8 == 1:
        print((n, x, "%8->", y))
    elif y % 4 == 1:
        print((n, x, "%4->", y))


(3, 49, '%8->', 9)
(7, 113, '%4->', 21)
(11, 177, '%16->', 33)
(15, 241, '%4->', 45)
(19, 305, '%8->', 57)
(23, 369, '%4->', 69)
(27, 433, '%16->', 81)
(31, 497, '%4->', 93)
(35, 561, '%8->', 105)
(39, 625, '%4->', 117)
(43, 689, '%16->', 129)
(47, 753, '%4->', 141)
(51, 817, '%8->', 153)
(55, 881, '%4->', 165)
(59, 945, '%16->', 177)
(63, 1009, '%4->', 189)
(67, 1073, '%8->', 201)
(71, 1137, '%4->', 213)
(75, 1201, '%16->', 225)
(79, 1265, '%4->', 237)


In [27]:
# 32n+1 numbers => 3n+1 numbers which are also 32m+1 numbers 1/32 of the time.
# case "011110...110111 "
for n in range(1, 160, 1):
    x = 32*n + 1
    y = (3*x + 1)//32
    if y % 32 == 1:
        print((n, x, "%32->", y))
    elif y % 16 == 1:
        print((n, x, "%16->", y))
    elif y % 8 == 1:
        print((n, x, "%8->", y))
    elif y % 4 == 1:
        print((n, x, "%4->", y))


(3, 97, '%8->', 9)
(7, 225, '%4->', 21)
(11, 353, '%32->', 33)
(15, 481, '%4->', 45)
(19, 609, '%8->', 57)
(23, 737, '%4->', 69)
(27, 865, '%16->', 81)
(31, 993, '%4->', 93)
(35, 1121, '%8->', 105)
(39, 1249, '%4->', 117)
(43, 1377, '%32->', 129)
(47, 1505, '%4->', 141)
(51, 1633, '%8->', 153)
(55, 1761, '%4->', 165)
(59, 1889, '%16->', 177)
(63, 2017, '%4->', 189)
(67, 2145, '%8->', 201)
(71, 2273, '%4->', 213)
(75, 2401, '%32->', 225)
(79, 2529, '%4->', 237)
(83, 2657, '%8->', 249)
(87, 2785, '%4->', 261)
(91, 2913, '%16->', 273)
(95, 3041, '%4->', 285)
(99, 3169, '%8->', 297)
(103, 3297, '%4->', 309)
(107, 3425, '%32->', 321)
(111, 3553, '%4->', 333)
(115, 3681, '%8->', 345)
(119, 3809, '%4->', 357)
(123, 3937, '%16->', 369)
(127, 4065, '%4->', 381)
(131, 4193, '%8->', 393)
(135, 4321, '%4->', 405)
(139, 4449, '%32->', 417)
(143, 4577, '%4->', 429)
(147, 4705, '%8->', 441)
(151, 4833, '%4->', 453)
(155, 4961, '%16->', 465)
(159, 5089, '%4->', 477)


In [28]:
# 4n+1 numbers => 3n+1 numbers which are also 4m+1 numbers 1/4 of the time.
# case "010...110111 "
for n in range(1, 160, 1):
    x = 4*n + 1
    y = (3*x + 1)//4
    if y % 32 == 1:
        print((n, x, "%32->", y))
    elif y % 16 == 1:
        print((n, x, "%16->", y))
    elif y % 8 == 1:
        print((n, x, "%8->", y))
    elif y % 4 == 1:
        print((n, x, "%4->", y))


(4, 17, '%4->', 13)
(8, 33, '%8->', 25)
(12, 49, '%4->', 37)
(16, 65, '%16->', 49)
(20, 81, '%4->', 61)
(24, 97, '%8->', 73)
(28, 113, '%4->', 85)
(32, 129, '%32->', 97)
(36, 145, '%4->', 109)
(40, 161, '%8->', 121)
(44, 177, '%4->', 133)
(48, 193, '%16->', 145)
(52, 209, '%4->', 157)
(56, 225, '%8->', 169)
(60, 241, '%4->', 181)
(64, 257, '%32->', 193)
(68, 273, '%4->', 205)
(72, 289, '%8->', 217)
(76, 305, '%4->', 229)
(80, 321, '%16->', 241)
(84, 337, '%4->', 253)
(88, 353, '%8->', 265)
(92, 369, '%4->', 277)
(96, 385, '%32->', 289)
(100, 401, '%4->', 301)
(104, 417, '%8->', 313)
(108, 433, '%4->', 325)
(112, 449, '%16->', 337)
(116, 465, '%4->', 349)
(120, 481, '%8->', 361)
(124, 497, '%4->', 373)
(128, 513, '%32->', 385)
(132, 529, '%4->', 397)
(136, 545, '%8->', 409)
(140, 561, '%4->', 421)
(144, 577, '%16->', 433)
(148, 593, '%4->', 445)
(152, 609, '%8->', 457)
(156, 625, '%4->', 469)


# So Lets make a graph of just the 4n+1 numbers that chain to 4n+1 numbers



In [55]:
def nextNode(n):
    chain = [n]
    m = n
    while m != 1:
        if m & 1 == 0:
            m = m // 2
        else:
            m = (3 * m + 1) // 2
        chain.append(m)
        if (m & 1):
            j = two_factors(m-1)
            if j>=2:
                if m < n:
                    print((n, m, "2^%d"%(j), chain))
                    break
                else:
                    print((n, m, "m > n", []))
#
    

In [57]:
for i in range(120):
    n = 4*i + 1
    nextNode(n)

(9, 17, 'm > n', [])
(9, 13, 'm > n', [])
(9, 5, '2^2', [9, 14, 7, 11, 17, 26, 13, 20, 10, 5])
(13, 5, '2^2', [13, 20, 10, 5])
(17, 13, '2^2', [17, 26, 13])
(25, 29, 'm > n', [])
(25, 17, '2^4', [25, 38, 19, 29, 44, 22, 11, 17])
(29, 17, '2^4', [29, 44, 22, 11, 17])
(33, 25, '2^3', [33, 50, 25])
(37, 17, '2^4', [37, 56, 28, 14, 7, 11, 17])
(41, 161, 'm > n', [])
(41, 121, 'm > n', [])
(41, 137, 'm > n', [])
(41, 233, 'm > n', [])
(41, 593, 'm > n', [])
(41, 445, 'm > n', [])
(41, 377, 'm > n', [])
(41, 425, 'm > n', [])
(41, 2429, 'm > n', [])
(41, 3077, 'm > n', [])
(41, 577, 'm > n', [])
(41, 433, 'm > n', [])
(41, 325, 'm > n', [])
(41, 61, 'm > n', [])
(41, 53, 'm > n', [])
(41, 5, '2^2', [41, 62, 31, 47, 71, 107, 161, 242, 121, 182, 91, 137, 206, 103, 155, 233, 350, 175, 263, 395, 593, 890, 445, 668, 334, 167, 251, 377, 566, 283, 425, 638, 319, 479, 719, 1079, 1619, 2429, 3644, 1822, 911, 1367, 2051, 3077, 4616, 2308, 1154, 577, 866, 433, 650, 325, 488, 244, 122, 61, 92, 46, 23, 3

In [48]:
nextNode(37)

(37, 17, '2^4', [37, 56, 28, 14, 7, 11, 17])


In [49]:
nextNode(61)

(61, 53, '2^2', [61, 92, 46, 23, 35, 53])


In [50]:
nextNode(53)

(53, 5, '2^2', [53, 80, 40, 20, 10, 5])


In [51]:
nextNode(25)

(25, 29, '2^2', [25, 38, 19, 29])


In [61]:
ChainPath(109), ChainPath(41), ChainPath(5)

('011010000010100100010000101100010010000001100001110101011101100011110111',
 '010000010100100010000101100010010000001100001110101011101100011110111',
 '0111')

# 4n+3 to 4n+1 sub-edges

Many more long runs, in general 4n+3 numbers tend to be farther from a 4n+1 number than another 4n+1 number.

In [62]:
for i in range(120):
    n = 4*i + 3
    nextNode(n)

(3, 5, 'm > n', [])
(7, 17, 'm > n', [])
(7, 13, 'm > n', [])
(7, 5, '2^2', [7, 11, 17, 26, 13, 20, 10, 5])
(11, 17, 'm > n', [])
(11, 13, 'm > n', [])
(11, 5, '2^2', [11, 17, 26, 13, 20, 10, 5])
(15, 53, 'm > n', [])
(15, 5, '2^2', [15, 23, 35, 53, 80, 40, 20, 10, 5])
(19, 29, 'm > n', [])
(19, 17, '2^4', [19, 29, 44, 22, 11, 17])
(23, 53, 'm > n', [])
(23, 5, '2^2', [23, 35, 53, 80, 40, 20, 10, 5])
(27, 41, 'm > n', [])
(27, 161, 'm > n', [])
(27, 121, 'm > n', [])
(27, 137, 'm > n', [])
(27, 233, 'm > n', [])
(27, 593, 'm > n', [])
(27, 445, 'm > n', [])
(27, 377, 'm > n', [])
(27, 425, 'm > n', [])
(27, 2429, 'm > n', [])
(27, 3077, 'm > n', [])
(27, 577, 'm > n', [])
(27, 433, 'm > n', [])
(27, 325, 'm > n', [])
(27, 61, 'm > n', [])
(27, 53, 'm > n', [])
(27, 5, '2^2', [27, 41, 62, 31, 47, 71, 107, 161, 242, 121, 182, 91, 137, 206, 103, 155, 233, 350, 175, 263, 395, 593, 890, 445, 668, 334, 167, 251, 377, 566, 283, 425, 638, 319, 479, 719, 1079, 1619, 2429, 3644, 1822, 911, 1367,

# 4n+1 to 4n+3 sub-edges

On the other hand, 4n+1 numbers generally lead to 4n+3 numbers quickly -- often the next step.


In [64]:
def nextNode_3(n):
    chain = [n]
    m = n
    while m != 1:
        if m & 1 == 0:
            m = m // 2
        else:
            m = (3 * m + 1) // 2
        chain.append(m)
        if m % 4 == 3:
            print((n, m, "3->1", chain))
            break
#
for i in range(120):
    n = 4*i + 3
    nextNode_3(n)

(7, 11, '3->1', [7, 11])
(15, 23, '3->1', [15, 23])
(19, 11, '3->1', [19, 29, 44, 22, 11])
(23, 35, '3->1', [23, 35])
(27, 31, '3->1', [27, 41, 62, 31])
(31, 47, '3->1', [31, 47])
(39, 59, '3->1', [39, 59])
(43, 7, '3->1', [43, 65, 98, 49, 74, 37, 56, 28, 14, 7])
(47, 71, '3->1', [47, 71])
(51, 11, '3->1', [51, 77, 116, 58, 29, 44, 22, 11])
(55, 83, '3->1', [55, 83])
(59, 67, '3->1', [59, 89, 134, 67])
(63, 95, '3->1', [63, 95])
(67, 19, '3->1', [67, 101, 152, 76, 38, 19])
(71, 107, '3->1', [71, 107])
(79, 119, '3->1', [79, 119])
(83, 47, '3->1', [83, 125, 188, 94, 47])
(87, 131, '3->1', [87, 131])
(91, 103, '3->1', [91, 137, 206, 103])
(95, 143, '3->1', [95, 143])
(99, 7, '3->1', [99, 149, 224, 112, 56, 28, 14, 7])
(103, 155, '3->1', [103, 155])
(107, 91, '3->1', [107, 161, 242, 121, 182, 91])
(111, 167, '3->1', [111, 167])
(115, 7, '3->1', [115, 173, 260, 130, 65, 98, 49, 74, 37, 56, 28, 14, 7])
(119, 179, '3->1', [119, 179])
(123, 139, '3->1', [123, 185, 278, 139])
(127, 191, '3->1'

In [110]:
FractionFromPath("010101")

(1, 1)

In [116]:
FractionFromPath("10101")

(2, 1)